README.md
LagerX
=======

This package implements a simple wrapper over https://github.com/basho/lager.

It embeds logging calls to LagerX into a module if currently configured logging
level is less or equal than severity of a call. Therefore it doesn't have
any negative impact on performance of a production system when you configure
error level even if you have tons of debug messages.

Information about location of a call (module, function, line, pid) is properly
passed to lager for your convenience so you can easily find the source of a message.
In this aspect using LagerX is equal to using parse transform shipped with
basho lager.

Since LagerX depends on macro implemented in LagerX module you have to require it.
Then you call one of logging methods on LagerX module. There are seven logging
methods in order of severity:

 - debug
 - info
 - notice
 - warning
 - error
 - critical
 - alert
 - emergency

Examples:
---------

```elixir
defmodule Test do
  require LagerX
  def debug do
    LagerX.debug "Hi debug"
  end
  def info do
    LagerX.info "Hi error"
  end
  def notice do
    LagerX.notice "Hi notice"
  end
  def warning do
    LagerX.warning "Hi warning"
  end
  def error do
    LagerX.error "Hi error"
  end
  def critical do
    LagerX.critical "Hi critical"
  end
  def alert do
    LagerX.alert "Hi alert"
  end
  def emergency do
    LagerX.emergency "Hi emergency"
  end
  def test do
    debug
    info
    notice
    warning
    error
    critical
    alert
    emergency
  end
end

Application.start :lager_x
Test.test
```

Configuration
-------------
It is possible to configure truncation size and compile time log level.
Being a simple wrapper LagerX doesn't attempt to configure underlying LagerX.
You would need to configure it yourself [see](https://github.com/basho/lager) to ensure that:

  * lager_truncation_size >= compile_truncation_size
  * lager severity level >= compile_log_level
  * appropriate handlers are configured

Configuration of LagerX can be done by calling helper functions of LagerX from your build system as follows:

```
iex(1)> LagerX.compile_log_level(:info)
true
iex(2)> LagerX.compile_truncation_size(512)
true
```

If you cannot call those function you can set compiler options:

```
iex(3)> Code.compiler_options lager_x_level: :debug
ok
iex(4)> Code.compiler_options lager_x_truncation_size: 512
ok
```

If you are mix user you could specify level and truncation_size in *config/config.#{Mix.env}.exs* as follows:

```
    use Mix.Config

    config :lager_x,
      level: :debug,
      truncation_size: 8096
```

Multiple Sink Support
---------------------
As of LagerX 3.x, you can configure multiple sinks to provide different behavior
for different streams of logs.  To use a different sink, prepend the name to the
logging calls above.  For example, to use the `magic_lager_event` sink, you can
do the following:

```
LagerX.info :magic, "magic event"
```


CHANGELOG.md
# v0.15.0 (2018-01-09)

* fork `:exlager`
* rename app name from `:exlager` to `:flager`
* rename main module from `Lager` to `Flager`
* set Elixir version to `~> 1.5`

# v0.3 (2013-04-21)

* enhancements
  * Use level name instead of meaningless integer

* deprecations
  * Use of integers to specify log level is depricated


test/flager_test.exs
defmodule Flager.Test do
  use ExUnit.Case
  @top Path.expand "../..", __ENV__.file

  @tag :potato
  test "debug" do
    {enabled, disabled} = split(compile(:debug))
    assert disabled == []
    assert enabled == [
     :alert, :critical, :debug, :emergency,
     :error, :info, :notice, :warning]
  end

  test "info" do
    {enabled, disabled} = split(compile(:info))
    assert disabled == [:debug]
    assert enabled == [
     :alert, :critical, :emergency,
     :error, :info, :notice, :warning]
  end

  test "notice" do
    {enabled, disabled} = split(compile(:notice))
    assert disabled == [:debug, :info]
    assert enabled == [:alert, :critical, :emergency, :error, :notice, :warning]
  end

  test "warning" do
    {enabled, disabled} = split(compile(:warning))
    assert disabled == [:debug, :info, :notice]
    assert enabled == [:alert, :critical, :emergency, :error, :warning]
  end

  test "error" do
    {enabled, disabled} = split(compile(:error))
    assert disabled == [:debug, :info, :notice, :warning]
    assert enabled == [:alert, :critical, :emergency, :error]
  end

  test "critical" do
    {enabled, disabled} = split(compile(:critical))
    assert disabled == [:debug, :error, :info, :notice, :warning]
    assert enabled == [:alert, :critical, :emergency]
  end

  test "alert" do
    {enabled, disabled} = split(compile(:alert))
    assert disabled == [:critical, :debug, :error, :info, :notice, :warning]
    assert enabled == [:alert, :emergency]
  end

  test "emergency" do
    {enabled, disabled} = split(compile(:emergency))
    assert disabled == [:alert, :critical, :debug, :error, :info, :notice, :warning]
    assert enabled == [:emergency]
  end

  test "none" do
    {enabled, disabled} = split(compile(:none))
    assert disabled == [:alert, :critical, :debug, :emergency,
      :error, :info, :notice, :warning]
    assert enabled == []
  end

  test "compile_log_level(atom)" do
    assert compile_log_level(:debug) == :debug
    assert compile_log_level(:info) == :info
    assert compile_log_level(:notice) == :notice
    assert compile_log_level(:warning) == :warning
    assert compile_log_level(:error) == :error
    assert compile_log_level(:critical) == :critical
    assert compile_log_level(:alert) == :alert
    assert compile_log_level(:emergency) == :emergency
    assert compile_log_level(:none) == :none
  end

  test "compile_log_level(integer)" do
    assert compile_log_level(7) == :debug
    assert compile_log_level(6) == :info
    assert compile_log_level(5) == :notice
    assert compile_log_level(4) == :warning
    assert compile_log_level(3) == :error
    assert compile_log_level(2) == :critical
    assert compile_log_level(1) == :alert
    assert compile_log_level(0) == :emergency
    assert compile_log_level(-1) == :none
  end

  test "get and set metadata" do
    assert [] = Flager.md
    new_md_list = [md1: "foo", md2: "bar"]
    assert :ok = Flager.md(new_md_list)
    assert ^new_md_list = Flager.md
  end

  setup_all do
    on_exit fn -> File.rm("#{@top}/test/#{beam(Flager)}") end
    :ok
  end

  defp compile_log_level(level) do
    true = Flager.compile_log_level(level)
    Flager.compile_log_level
  end

  defp compile(level) do
    :code.purge Flager
    Application.put_env :flager, :level, level
    Kernel.ParallelCompiler.files_to_path ["#{@top}/lib/flager.ex"], "#{@top}/test"
    Code.ensure_compiled(Flager)
    quoted =
      quote do
        require Flager
        [
         debug: Flager.debug("Hi debug"),
         info: Flager.info("Hi info"),
         notice: Flager.notice("Hi notice"),
         warning: Flager.warning("Hi warning"),
         error: Flager.error("Hi error"),
         critical: Flager.critical("Hi critical"),
         alert: Flager.alert("Hi alert"),
         emergency: Flager.emergency("Hi emergency"),
        ]
      end
    {res, _} = Code.eval_quoted quoted
    res
  end

  defp beam(module), do: "#{module}.beam"

  defp split(macros) do
    {e, d} = Enum.reduce macros, {[], []}, fn({level, res}, {e, d}) ->
      if is_nil(res) do
        {e, [level|d]}
      else
        {[level|e], d}
      end
    end
    {Enum.sort(e), Enum.sort(d)}
  end
end


test/exlager_test.exs
defmodule Flager.Test do
  use ExUnit.Case
  @top Path.expand "../..", __ENV__.file

  test "debug" do
    {enabled, disabled} = split(compile(:debug))
    assert disabled == []
    assert enabled == [
     :alert, :critical, :debug, :emergency,
     :error, :info, :notice, :warning]
  end

  test "info" do
    {enabled, disabled} = split(compile(:info))
    assert disabled == [:debug]
    assert enabled == [
     :alert, :critical, :emergency,
     :error, :info, :notice, :warning]
  end

  test "notice" do
    {enabled, disabled} = split(compile(:notice))
    assert disabled == [:debug, :info]
    assert enabled == [:alert, :critical, :emergency, :error, :notice, :warning]
  end

  test "warning" do
    {enabled, disabled} = split(compile(:warning))
    assert disabled == [:debug, :info, :notice]
    assert enabled == [:alert, :critical, :emergency, :error, :warning]
  end

  test "error" do
    {enabled, disabled} = split(compile(:error))
    assert disabled == [:debug, :info, :notice, :warning]
    assert enabled == [:alert, :critical, :emergency, :error]
  end

  test "critical" do
    {enabled, disabled} = split(compile(:critical))
    assert disabled == [:debug, :error, :info, :notice, :warning]
    assert enabled == [:alert, :critical, :emergency]
  end

  test "alert" do
    {enabled, disabled} = split(compile(:alert))
    assert disabled == [:critical, :debug, :error, :info, :notice, :warning]
    assert enabled == [:alert, :emergency]
  end

  test "emergency" do
    {enabled, disabled} = split(compile(:emergency))
    assert disabled == [:alert, :critical, :debug, :error, :info, :notice, :warning]
    assert enabled == [:emergency]
  end

  test "none" do
    {enabled, disabled} = split(compile(:none))
    assert disabled == [:alert, :critical, :debug, :emergency,
      :error, :info, :notice, :warning]
    assert enabled == []
  end

  test "compile_log_level(atom)" do
    assert compile_log_level(:debug) == :debug
    assert compile_log_level(:info) == :info
    assert compile_log_level(:notice) == :notice
    assert compile_log_level(:warning) == :warning
    assert compile_log_level(:error) == :error
    assert compile_log_level(:critical) == :critical
    assert compile_log_level(:alert) == :alert
    assert compile_log_level(:emergency) == :emergency
    assert compile_log_level(:none) == :none
  end

  test "compile_log_level(integer)" do
    assert compile_log_level(7) == :debug
    assert compile_log_level(6) == :info
    assert compile_log_level(5) == :notice
    assert compile_log_level(4) == :warning
    assert compile_log_level(3) == :error
    assert compile_log_level(2) == :critical
    assert compile_log_level(1) == :alert
    assert compile_log_level(0) == :emergency
    assert compile_log_level(-1) == :none
  end

  test "get and set metadata" do
    assert [] = Lager.md
    new_md_list = [md1: "foo", md2: "bar"]
    assert :ok = Lager.md(new_md_list)
    assert ^new_md_list = Lager.md
  end

  setup_all do
    on_exit fn -> File.rm("#{@top}/test/#{beam(Lager)}") end
    :ok
  end

  defp compile_log_level(level) do
    true = Lager.compile_log_level(level)
    Lager.compile_log_level
  end

  defp compile(level) do
    :code.purge Lager
    Application.put_env :flager, :level, level
    Kernel.ParallelCompiler.files_to_path ["#{@top}/lib/lager.ex"], "#{@top}/test"
    Code.ensure_compiled(Lager)
    quoted =
      quote do
        require Lager
        [
         debug: Lager.debug("Hi debug"),
         info: Lager.info("Hi info"),
         notice: Lager.notice("Hi notice"),
         warning: Lager.warning("Hi warning"),
         error: Lager.error("Hi error"),
         critical: Lager.critical("Hi critical"),
         alert: Lager.alert("Hi alert"),
         emergency: Lager.emergency("Hi emergency"),
        ]
      end
    {res, _} = Code.eval_quoted quoted
    res
  end

  defp beam(module), do: "#{module}.beam"

  defp split(macros) do
    {e, d} = Enum.reduce macros, {[], []}, fn({level, res}, {e, d}) ->
      if is_nil(res) do
        {e, [level|d]}
      else
        {[level|e], d}
      end
    end
    {Enum.sort(e), Enum.sort(d)}
  end
end


test/lager_x_test.exs
defmodule LagerX.Test do
  use ExUnit.Case
  @top Path.expand "../..", __ENV__.file

  test "debug" do
    {enabled, disabled} = split(compile(:debug))
    assert disabled == []
    assert enabled == [
     :alert, :critical, :debug, :emergency,
     :error, :info, :notice, :warning]
  end

  test "info" do
    {enabled, disabled} = split(compile(:info))
    assert disabled == [:debug]
    assert enabled == [
     :alert, :critical, :emergency,
     :error, :info, :notice, :warning]
  end

  test "notice" do
    {enabled, disabled} = split(compile(:notice))
    assert disabled == [:debug, :info]
    assert enabled == [:alert, :critical, :emergency, :error, :notice, :warning]
  end

  test "warning" do
    {enabled, disabled} = split(compile(:warning))
    assert disabled == [:debug, :info, :notice]
    assert enabled == [:alert, :critical, :emergency, :error, :warning]
  end

  test "error" do
    {enabled, disabled} = split(compile(:error))
    assert disabled == [:debug, :info, :notice, :warning]
    assert enabled == [:alert, :critical, :emergency, :error]
  end

  test "critical" do
    {enabled, disabled} = split(compile(:critical))
    assert disabled == [:debug, :error, :info, :notice, :warning]
    assert enabled == [:alert, :critical, :emergency]
  end

  test "alert" do
    {enabled, disabled} = split(compile(:alert))
    assert disabled == [:critical, :debug, :error, :info, :notice, :warning]
    assert enabled == [:alert, :emergency]
  end

  test "emergency" do
    {enabled, disabled} = split(compile(:emergency))
    assert disabled == [:alert, :critical, :debug, :error, :info, :notice, :warning]
    assert enabled == [:emergency]
  end

  test "none" do
    {enabled, disabled} = split(compile(:none))
    assert disabled == [:alert, :critical, :debug, :emergency,
      :error, :info, :notice, :warning]
    assert enabled == []
  end

  test "compile_log_level(atom)" do
    assert compile_log_level(:debug) == :debug
    assert compile_log_level(:info) == :info
    assert compile_log_level(:notice) == :notice
    assert compile_log_level(:warning) == :warning
    assert compile_log_level(:error) == :error
    assert compile_log_level(:critical) == :critical
    assert compile_log_level(:alert) == :alert
    assert compile_log_level(:emergency) == :emergency
    assert compile_log_level(:none) == :none
  end

  test "compile_log_level(integer)" do
    assert compile_log_level(7) == :debug
    assert compile_log_level(6) == :info
    assert compile_log_level(5) == :notice
    assert compile_log_level(4) == :warning
    assert compile_log_level(3) == :error
    assert compile_log_level(2) == :critical
    assert compile_log_level(1) == :alert
    assert compile_log_level(0) == :emergency
    assert compile_log_level(-1) == :none
  end

  test "get and set metadata" do
    assert [] = LagerX.md
    new_md_list = [md1: "foo", md2: "bar"]
    assert :ok = LagerX.md(new_md_list)
    assert ^new_md_list = LagerX.md
  end

  setup_all do
    on_exit fn -> File.rm("#{@top}/test/#{beam(LagerX)}") end
    :ok
  end

  defp compile_log_level(level) do
    true = LagerX.compile_log_level(level)
    LagerX.compile_log_level
  end

  defp compile(level) do
    :code.purge LagerX
    Application.put_env :lager_x, :level, level
    Kernel.ParallelCompiler.files_to_path ["#{@top}/lib/lager_x.ex"], "#{@top}/test"
    Code.ensure_compiled(LagerX)
    quoted =
      quote do
        require LagerX
        [
         debug: LagerX.debug("Hi debug"),
         info: LagerX.info("Hi info"),
         notice: LagerX.notice("Hi notice"),
         warning: LagerX.warning("Hi warning"),
         error: LagerX.error("Hi error"),
         critical: LagerX.critical("Hi critical"),
         alert: LagerX.alert("Hi alert"),
         emergency: LagerX.emergency("Hi emergency"),
        ]
      end
    {res, _} = Code.eval_quoted quoted
    res
  end

  defp beam(module), do: "#{module}.beam"

  defp split(macros) do
    {e, d} = Enum.reduce macros, {[], []}, fn({level, res}, {e, d}) ->
      if is_nil(res) do
        {e, [level|d]}
      else
        {[level|e], d}
      end
    end
    {Enum.sort(e), Enum.sort(d)}
  end
end


lib/lager.ex
defmodule Lager do
  @default_sink :lager_event

  defdelegate trace_console(filter), to: :lager
  defdelegate trace_file(file, filter, level), to: :lager
  defdelegate stop_trace(trace), to: :lager
  defdelegate clear_all_traces(), to: :lager
  defdelegate status(), to: :lager
  defdelegate set_loglevel(handler, level), to: :lager
  defdelegate set_loglevel(handler, indent, level), to: :lager
  defdelegate get_loglevel(handler), to: :lager
  defdelegate posix_error(error), to: :lager
  defdelegate md, to: :lager
  defdelegate md(new_md_list), to: :lager

  levels = [
    debug:      7,
    info:       6,
    notice:     5,
    warning:    4,
    error:      3,
    critical:   2,
    alert:      1,
    emergency:  0,
    none:      -1
  ]

  quoted = for {level, _num} <- levels do
    quote do
      defmacro unquote(level)(message) do
        log(@default_sink, unquote(level), '~ts', [message], __CALLER__)
      end
      defmacro unquote(level)(sink, message) when is_atom(sink) do
        sink_evt = String.to_atom(Atom.to_string(sink) <> "_lager_event")
        log(sink_evt, unquote(level), '~ts', [message], __CALLER__)
      end
      defmacro unquote(level)(format, message) do
        log(@default_sink, unquote(level), format, message, __CALLER__)
      end
      defmacro unquote(level)(sink, format, message) when is_atom(sink) do
        sink_evt = String.to_atom(Atom.to_string(sink) <> "_lager_event")
        log(sink_evt, unquote(level), format, message, __CALLER__)
      end
    end
  end
  Module.eval_quoted __MODULE__, quoted, [], __ENV__

  quoted = for {level, num} <- levels do
    quote do
      defp level_to_num(unquote(level)), do: unquote(num)
    end
  end
  Module.eval_quoted __MODULE__, quoted, [], __ENV__
  defp level_to_num(_), do: nil

  quoted = for {level, num} <- levels do
    quote do
      defp num_to_level(unquote(num)), do:  unquote(level)
    end
  end
  Module.eval_quoted __MODULE__, quoted, [], __ENV__
  defp num_to_level(_), do: nil

  defp log(sink, level, format, args, caller) do
    {name, _arity} = caller.function || {:unknown, 0}
    module = caller.module || :unknown
    format =
      if is_binary(format) do
        String.to_charlist(format)
      else
        format
      end
    if should_log(level) do
      dispatch(sink, level, module, name, caller.line, format, args)
    end
  end

  defp dispatch(sink, level, module, name, line, format, args) do
    quote do
      :lager.dispatch_log(unquote(sink), unquote(level),
        [module: unquote(module),
         function: unquote(name),
         line: unquote(line),
         pid: self(),
         node: node()] ++ :lager.md(),
        unquote(format), unquote(args), unquote(compile_truncation_size()),
        :safe)
    end
  end

  defp should_log(level), do: level_to_num(level) <= level_to_num(compile_log_level())

  @doc """
  This function is used to get compile time log level.
  Examples:
    iex(4)> Lager.compile_log_level
    :info
  """
  def compile_log_level() do
    level = Application.get_env(:flager, :level, :info)
    if is_integer(level) do
      level = num_to_level(level)
      IO.puts "Using integers is deprecated, please use :#{level} instead"
      level
    else
      level
    end
  end

  @doc """
  This function is used to set compile time log level.
  By default the log level is 'info'.
  Examples:
    iex(4)> Lager.compile_log_level(6)
    true
    iex(4)> Lager.compile_log_level(:info)
    true
  """
  def compile_log_level(level) when level in -1..7 do
    compile_log_level(num_to_level(level))
  end
  def compile_log_level(level) when is_atom(level) do
    :ok = Application.put_env(:flager, :level, level)
    true
  end
  def compile_log_level(level) do
    IO.puts "ERROR: unknown level #{inspect level}"
    false
  end

  def compile_truncation_size() do
    Application.get_env(:flager, :truncation_size, 4096)
  end

  @doc """
  This function is used to set compile time truncation size.
  By default the truncation size is 4096.
  Examples:
    iex(4)> Lager.compile_truncation_size(512)
    true
  """
  def compile_truncation_size(size) do
    :ok = Application.put_env(:flager, :truncation_size, size)
    true
  end
end


lib/flager.ex
defmodule Flager do
  @default_sink :lager_event

  defdelegate trace_console(filter), to: :lager
  defdelegate trace_file(file, filter, level), to: :lager
  defdelegate stop_trace(trace), to: :lager
  defdelegate clear_all_traces(), to: :lager
  defdelegate status(), to: :lager
  defdelegate set_loglevel(handler, level), to: :lager
  defdelegate set_loglevel(handler, indent, level), to: :lager
  defdelegate get_loglevel(handler), to: :lager
  defdelegate posix_error(error), to: :lager
  defdelegate md, to: :lager
  defdelegate md(new_md_list), to: :lager

  levels = [
    debug:      7,
    info:       6,
    notice:     5,
    warning:    4,
    error:      3,
    critical:   2,
    alert:      1,
    emergency:  0,
    none:      -1
  ]

  quoted = for {level, _num} <- levels do
    quote do
      defmacro unquote(level)(message) do
        log(@default_sink, unquote(level), '~ts', [message], __CALLER__)
      end
      defmacro unquote(level)(sink, message) when is_atom(sink) do
        sink_evt = String.to_atom(Atom.to_string(sink) <> "_lager_event")
        log(sink_evt, unquote(level), '~ts', [message], __CALLER__)
      end
      defmacro unquote(level)(format, message) do
        log(@default_sink, unquote(level), format, message, __CALLER__)
      end
      defmacro unquote(level)(sink, format, message) when is_atom(sink) do
        sink_evt = String.to_atom(Atom.to_string(sink) <> "_lager_event")
        log(sink_evt, unquote(level), format, message, __CALLER__)
      end
    end
  end
  Module.eval_quoted __MODULE__, quoted, [], __ENV__

  quoted = for {level, num} <- levels do
    quote do
      defp level_to_num(unquote(level)), do: unquote(num)
    end
  end
  Module.eval_quoted __MODULE__, quoted, [], __ENV__
  defp level_to_num(_), do: nil

  quoted = for {level, num} <- levels do
    quote do
      defp num_to_level(unquote(num)), do:  unquote(level)
    end
  end
  Module.eval_quoted __MODULE__, quoted, [], __ENV__
  defp num_to_level(_), do: nil

  defp log(sink, level, format, args, caller) do
    {name, _arity} = caller.function || {:unknown, 0}
    module = caller.module || :unknown
    format =
      if is_binary(format) do
        String.to_charlist(format)
      else
        format
      end
    if should_log(level) do
      dispatch(sink, level, module, name, caller.line, format, args)
    end
  end

  defp dispatch(sink, level, module, name, line, format, args) do
    quote do
      :lager.dispatch_log(unquote(sink), unquote(level),
        [module: unquote(module),
         function: unquote(name),
         line: unquote(line),
         pid: self(),
         node: node()] ++ :lager.md(),
        unquote(format), unquote(args), unquote(compile_truncation_size()),
        :safe)
    end
  end

  defp should_log(level), do: level_to_num(level) <= level_to_num(compile_log_level())

  @doc """
  This function is used to get compile time log level.
  Examples:
    iex(4)> Flager.compile_log_level
    :info
  """
  def compile_log_level() do
    level = Application.get_env(:flager, :level, :info)
    if is_integer(level) do
      level = num_to_level(level)
      IO.puts "Using integers is deprecated, please use :#{level} instead"
      level
    else
      level
    end
  end

  @doc """
  This function is used to set compile time log level.
  By default the log level is 'info'.
  Examples:
    iex(4)> Flager.compile_log_level(6)
    true
    iex(4)> Flager.compile_log_level(:info)
    true
  """
  def compile_log_level(level) when level in -1..7 do
    compile_log_level(num_to_level(level))
  end
  def compile_log_level(level) when is_atom(level) do
    :ok = Application.put_env(:flager, :level, level)
    true
  end
  def compile_log_level(level) do
    IO.puts "ERROR: unknown level #{inspect level}"
    false
  end

  def compile_truncation_size() do
    Application.get_env(:flager, :truncation_size, 4096)
  end

  @doc """
  This function is used to set compile time truncation size.
  By default the truncation size is 4096.
  Examples:
    iex(4)> Flager.compile_truncation_size(512)
    true
  """
  def compile_truncation_size(size) do
    :ok = Application.put_env(:flager, :truncation_size, size)
    true
  end
end


lib/lager_x.ex
defmodule LagerX do
  @default_sink :lager_event

  defdelegate trace_console(filter), to: :lager
  defdelegate trace_file(file, filter, level), to: :lager
  defdelegate stop_trace(trace), to: :lager
  defdelegate clear_all_traces(), to: :lager
  defdelegate status(), to: :lager
  defdelegate set_loglevel(handler, level), to: :lager
  defdelegate set_loglevel(handler, indent, level), to: :lager
  defdelegate get_loglevel(handler), to: :lager
  defdelegate posix_error(error), to: :lager
  defdelegate md, to: :lager
  defdelegate md(new_md_list), to: :lager

  levels = [
    debug:      7,
    info:       6,
    notice:     5,
    warning:    4,
    error:      3,
    critical:   2,
    alert:      1,
    emergency:  0,
    none:      -1
  ]

  quoted = for {level, _num} <- levels do
    quote do
      defmacro unquote(level)(message) do
        log(@default_sink, unquote(level), '~ts', [message], __CALLER__)
      end
      defmacro unquote(level)(sink, message) when is_atom(sink) do
        sink_evt = String.to_atom(Atom.to_string(sink) <> "_lager_event")
        log(sink_evt, unquote(level), '~ts', [message], __CALLER__)
      end
      defmacro unquote(level)(format, message) do
        log(@default_sink, unquote(level), format, message, __CALLER__)
      end
      defmacro unquote(level)(sink, format, message) when is_atom(sink) do
        sink_evt = String.to_atom(Atom.to_string(sink) <> "_lager_event")
        log(sink_evt, unquote(level), format, message, __CALLER__)
      end
    end
  end
  Module.eval_quoted __MODULE__, quoted, [], __ENV__

  quoted = for {level, num} <- levels do
    quote do
      defp level_to_num(unquote(level)), do: unquote(num)
    end
  end
  Module.eval_quoted __MODULE__, quoted, [], __ENV__
  defp level_to_num(_), do: nil

  quoted = for {level, num} <- levels do
    quote do
      defp num_to_level(unquote(num)), do:  unquote(level)
    end
  end
  Module.eval_quoted __MODULE__, quoted, [], __ENV__
  defp num_to_level(_), do: nil

  defp log(sink, level, format, args, caller) do
    {name, _arity} = caller.function || {:unknown, 0}
    module = caller.module || :unknown
    format =
      if is_binary(format) do
        String.to_char_list(format)
      else
        format
      end
    if should_log(level) do
      dispatch(sink, level, module, name, caller.line, format, args)
    end
  end

  defp dispatch(sink, level, module, name, line, format, args) do
    quote do
      :lager.dispatch_log(unquote(sink), unquote(level),
        [module: unquote(module),
         function: unquote(name),
         line: unquote(line),
         pid: self(),
         node: node()] ++ :lager.md(),
        unquote(format), unquote(args), unquote(compile_truncation_size()),
        :safe)
    end
  end

  defp should_log(level), do: level_to_num(level) <= level_to_num(compile_log_level())

  @doc """
  This function is used to get compile time log level.
  Examples:
    iex(4)> LagerX.compile_log_level
    :info
  """
  def compile_log_level() do
    level = Application.get_env(:lager_x, :level, :info)
    if is_integer(level) do
      level = num_to_level(level)
      IO.puts "Using integers is deprecated, please use :#{level} instead"
      level
    else
      level
    end
  end

  @doc """
  This function is used to set compile time log level.
  By default the log level is 'info'.
  Examples:
    iex(4)> LagerX.compile_log_level(6)
    true
    iex(4)> LagerX.compile_log_level(:info)
    true
  """
  def compile_log_level(level) when level in -1..7 do
    compile_log_level(num_to_level(level))
  end
  def compile_log_level(level) when is_atom(level) do
    :ok = Application.put_env(:lager_x, :level, level)
    true
  end
  def compile_log_level(level) do
    IO.puts "ERROR: unknown level #{inspect level}"
    false
  end

  def compile_truncation_size() do
    Application.get_env(:lager_x, :truncation_size, 4096)
  end

  @doc """
  This function is used to set compile time truncation size.
  By default the truncation size is 4096.
  Examples:
    iex(4)> LagerX.compile_truncation_size(512)
    true
  """
  def compile_truncation_size(size) do
    :ok = Application.put_env(:lager_x, :truncation_size, size)
    true
  end
end
